### PR TITLE
Handle simultanous open in Handshake

### PR DESCRIPTION
**Pull request by coot**
- handshake codec
- typed-protocols: added TokPeerRole
- handshake: resiliant for simultanous open
- handshake: updated `network-spec`

### CODE REVIEW

File: typed-protocols/src/Network/TypedProtocol/Core.hs [hunk 1]
```diff
@@ -320,6 +322,14 @@ class Protocol ps where
 --
 data PeerRole = AsClient | AsServer
 
+-- | Singletons for the promoted 'PeerRole' types.  Not directly used be the
```
[the remark this answers is not in the document]
Comment (coot) [reply typed-protocols/src/Network/TypedProtocol/Core.hs]: Thanks, done

**Comment by karknu** [comment]
Given that this fix triggers race conditions  between governor and peer state actions I don't think this PR should be merged before #3167 .

File: ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Codec.hs [hunk 2]
```diff
@@ -141,6 +141,19 @@ codecHandshake versionNumberCodec = mkCodecCborLazyBS encodeMsg decodeMsg
...
+        let vs' = Map.toAscList vs
+        in
+           CBOR.encodeListLen 2
+        <> CBOR.encodeWord 0
+        <> CBOR.encodeMapLen (fromIntegral $ length vs')
+        <> mconcat [    CBOR.encodeTerm (encodeTerm versionNumberCodec vNumber)
+                     <> CBOR.encodeTerm vParams
+                   | (vNumber, vParams) <- vs'
+                   ]
+
```
Comment (karknu) [inline ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Codec.hs:156]: This seems error prone, it is just a duplication of the encoder for `MsgProposeVersions` . Can you combine them?

File: ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Type.hs [hunk 3]
```diff
@@ -34,16 +35,15 @@ import           Data.Map (Map)
...
 -- associated parameters of the protocol to use for all subsequent
 -- communication.
 --
 data Handshake vNumber vParams where
-    StPropose :: Handshake vNumber vParams
-    StConfirm :: Handshake vNumber vParams
-    StDone    :: Handshake vNumber vParams
+    StPropose  :: Handshake vNumber vParams
+    StConfirm  :: Handshake vNumber vParams
+    StDone     :: Handshake vNumber vParams
```
Comment (karknu) [inline ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Type.hs:46]: I don't see the point of the extra space, the `::` where aligned.

File: ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Type.hs [hunk 4]
```diff
@@ -111,11 +121,28 @@ instance Protocol (Handshake vNumber vParams) where
     exclusionLemma_NobodyAndClientHaveAgency TokDone    tok = case tok of {}
     exclusionLemma_NobodyAndServerHaveAgency TokDone    tok = case tok of {}
 
+-- $simultanous-open
+--
+-- On simultanous open both sides will send `MsgProposeVersions`, which will be
+-- decoded as `MsgProposeVersions'` and both sides will reply with either
+-- 'MsgAcceptVersion' or 'MsgRefuse'.  It is important to stress that in this
+-- case both sides will make the choice which version and parameters to pick.
```
Comment (karknu) [inline ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Type.hs:129]: This isn't correct anymore. Both sides will send `MsgProposeVersions` , decode their received `MsgProposeVersions'` message and if both sides agree no further handshake messages will be sent.

File: ouroboros-network/protocol-tests/Ouroboros/Network/Protocol/Handshake/Test.hs [hunk 5]
```diff
@@ -539,6 +545,103 @@ prop_pipe_asymmetric_IO (ArbitraryVersions clientVersions _serverVersions) =
...
+  -                 clientVersions')) of
+  -       (rClientRes, rClientRes', TerminalStates TokDone TokDone) ->
+  -            fromMaybe False eClientRes
+  -            ===
+  -            either (const False) (\(a,_,_) -> a) rClientRes
+  -         .&&.
+  -            fromMaybe False eClientRes'
+  -            ===
+  -            either (const False) (\(a,_,_) -> a) rClientRes'
+  --}
```
Comment (karknu) [inline ouroboros-network/protocol-tests/Ouroboros/Network/Protocol/Handshake/Test.hs:576]: Why do you need to commit this commented out code?
If it is needed can you add a comment on why it is worth keeping around.

File: docs/network-spec/miniprotocols.tex [hunk 6]
```diff
@@ -376,44 +377,54 @@ \subsection{Description}
 supported protocol versions and callback functions for handling the protocol parameters.
 These callback functions are specific for the supported protocol versions.
 
+The handshake mini protocol can is designed to handle simultantous TCP open.
```
Comment (karknu) [inline docs/network-spec/miniprotocols.tex:380]: s/can is/is/

**Review by karknu — CHANGES_REQUESTED** [review]
Looks good. Just some minor comments.

File: ouroboros-network/protocol-tests/Ouroboros/Network/Protocol/Handshake/Test.hs (hunk 5 above)
Comment (coot) [reply ouroboros-network/protocol-tests/Ouroboros/Network/Protocol/Handshake/Test.hs:576]: This should be removed.

**Review by karknu — APPROVED** [review]
LGTM

**Comment by coot** [comment]
bors merge

**Pull request merged**